File: ios/LofiGlass/Services/LofiFilter.swift
import Foundation

// MARK: - The lofi gate
//
// "finds only youtube lofi stuff" is a filter problem, not a search problem:
// YouTube returns jazz-café ambience and deep-house festival mixes for the same
// queries. So every candidate — from the seed corpus or a live search — is
// scored on title + description + hashtags + channel + length, and anything
// under the threshold is dropped (and remembered, so the UI can show what it
// threw away).
//
// Weights and threshold are the same ones in prototype/src/lofi-filter.js, and
// `node tools/gate-check.mjs` is the executable reference for both.

enum LofiFilter {
  static let threshold: Double = 4

  private struct Signal {
    let pattern: String
    let weight: Double
    let tag: String
  }

  private static let positive: [Signal] = [
    .init(pattern: "lo[.\\-\\s]?fi", weight: 5, tag: "lofi"),
    .init(pattern: "l o f i", weight: 4, tag: "lofi"),
    .init(pattern: "\u{FF4C}\u{FF4F}\u{FF46}\u{FF49}", weight: 4, tag: "lofi"),
    .init(pattern: "chill ?hop", weight: 4, tag: "chillhop"),
    .init(pattern: "jazz ?hop", weight: 4, tag: "jazzy"),
    .init(pattern: "(music|beats|sounds|tunes|radio)\\s*(to|for)\\s+(relax|study|sleep|focus|chill|work|drive|code)", weight: 3.5, tag: "beats to \u{2026}"),
    .init(pattern: "beats? ?to ?(relax|study|sleep|chill|focus|drive)", weight: 4, tag: "beats to \u{2026}"),
    .init(pattern: "study beats", weight: 3, tag: "study"),
    .init(pattern: "instrumental hip ?hop", weight: 3, tag: "instrumental"),
    .init(pattern: "hip ?hop radio", weight: 3, tag: "radio"),
    .init(pattern: "boom ?bap", weight: 3, tag: "boom bap"),
    .init(pattern: "tape hiss|tape loop|4th ?gen tape", weight: 2.5, tag: "tape"),
    .init(pattern: "\\bvhs\\b|crt|scan ?line", weight: 2, tag: "vhs"),
    .init(pattern: "chill beats", weight: 2, tag: "chill"),
    .init(pattern: "sleep lofi", weight: 3, tag: "sleep"),
    .init(pattern: "anime (edit|loop|amv)", weight: 2, tag: "anime edit"),
    .init(pattern: "type beat", weight: 2, tag: "type beat"),
    .init(pattern: "neo soul", weight: 1.5, tag: "neo soul"),
    .init(pattern: "\\bjazzy\\b", weight: 1.5, tag: "jazzy"),
    .init(pattern: "24\\s*/\\s*7", weight: 2, tag: "24/7"),
    .init(pattern: "lofi girl|chilledcow", weight: 3, tag: "lofi girl"),
    .init(pattern: "\\b(?:prod\\.|beat)\\s*(?:by\\s*)?[a-z0-9_.-]+\\b", weight: 1, tag: "beatmaker"),
  ]

  private static let negative: [Signal] = [
    .init(pattern: "deep house", weight: -4, tag: "deep house"),
    .init(pattern: "progressive house", weight: -4, tag: "progressive house"),
    .init(pattern: "\u{0008}trance\u{0008}", weight: -3.5, tag: "trance"),
    .init(pattern: "hard ?style", weight: -5, tag: "hardstyle"),
    .init(pattern: "phonk", weight: -3, tag: "phonk"),
    .init(pattern: "\u{0008}drill\u{0008}", weight: -2, tag: "drill"),
    .init(pattern: "heavy metal|deathcore|metalcore", weight: -5, tag: "metal"),
    .init(pattern: "k[s-]?pop", weight: -2, tag: "k-pop"),
    .init(pattern: "binaural", weight: -3, tag: "binaural beats"),
    .init(pattern: "subliminal|affirmation|solfeggio|d{3} ?hz", weight: -3.5, tag: "frequencies/affirmations"),
    .init(pattern: "karaoke", weight: -3, tag: "karaoke"),
    .init(pattern: "workout|gym motivation", weight: -3, tag: "workout"),
    .init(pattern: "edm (festival|mix)", weight: -4, tag: "edm"),
    .init(pattern: "lofi (fake|filter scam)", weight: -5, tag: "lofi-bait"),
  ]

  /// Channels whose whole catalogue is curated lofi.
  static let trustedChannels: [String] = [
    "lofi girl", "chilledcow", "lofi records", "chillhop music",
    "settle", "the bootleg boy", "afro lofi", "jeez",
    "lofi coffee", "mimi lofi chill", "the japanese town", "a lofi soul",
    "lofi shop 24h", "flux.fm", "lofi corners"
  ]

  /// Producers who only exist in this scene — a bare "Artist - Title" upload
  /// from one of them is lofi even with an empty description.
  static let knownActs: [String] = [
    "kudasai", "no spirit", "tonion", "nymano",
    "yasumu", "hm surf", "lilac", "trxxshed",
    "jhove", "blurred figures", "another silent weekend", "swiftly",
    "hazue", "noji", "sutton", "thymes",
    "home grown", "luella", "idealism", "sleep bean",
    "jinsang", "fantompower", "powfu", "dj hazel",
    "luvlee", "mnts", "tenncoats", "cwrd",
    "philo", "kuun", "vanyforce", "dthcheese",
    "shimza", "purrple cat"
  ]

  private static let moods: [(name: String, words: [String])] = [
    ("sleep", ["sleep", "bedtime", "dream", "insomnia", "8 hours", "\u{1F4A4}"]),
    ("study", ["study", "exam", "homework", "revision", "relax/study"]),
    ("rain", ["rain", "storm", "thunder", "wet"]),
    ("cafe", ["cafe", "coffee", "barista", "latte"]),
    ("night-drive", ["night drive", "headlights", "midnight", "\u{FF5E} drive", "drive to"]),
    ("jazzy", ["jazz", "sax", "piano trio", "swing"]),
    ("sad", ["sad", "cry", "lonely", "heartbreak", "melanch", "\u{1F494}"]),
    ("anime", ["anime", "ghibli", "spirited away", "opening"]),
    ("morning", ["morning", "sunrise", "breakfast", "upbeat"]),
    ("code", ["code", "coding", "programming"]),
    ("focus", ["focus", "deep work", "concentration"]),
  ]

  /// Surrounds with spaces and squashes separators, so word-boundary checks can
  /// be plain `contains` without fighting regex escaping.
  static func normalized(_ s: String) -> String {
    let lowered = s.lowercased()
    var out = " "
    for ch in lowered {
      if ch == "|" || ch == "·" || ch == "—" || ch == "–" || ch == "\n" || ch == "\r" || ch == "\t" { out.append(" ") }
      else { out.append(ch) }
    }
    while out.contains("  ") { out = out.replacingOccurrences(of: "  ", with: " ") }
    return out + " "
  }

  private static func matches(_ pattern: String, in text: String) -> Bool {
    text.range(of: pattern, options: [.regularExpression, .caseInsensitive]) != nil
  }

  static func evaluate(
    title: String,
    description: String = "",
    tags: [String] = [],
    channelName: String = "",
    durationSeconds: Int = 0
  ) -> GateVerdict {
    let titleHay = normalized(title)
    let hay = titleHay + "|" + normalized(description) + "|" + normalized(tags.joined(separator: " ")) + "|" + normalized(channelName)

    var score: Double = 0
    var hits: [String] = []
    var penalties: [String] = []

    for signal in positive {
      let inTitle = matches(signal.pattern, in: titleHay)
      if inTitle || matches(signal.pattern, in: hay) {
        score += signal.weight + (inTitle ? 1 : 0)
        if !hits.contains(signal.tag) { hits.append(signal.tag) }
      }
    }

    for signal in negative where matches(signal.pattern, in: hay) {
      score += signal.weight
      if !penalties.contains(signal.tag) { penalties.append(signal.tag) }
    }

    let channelKey = normalized(channelName)
    if trustedChannels.contains(where: { channelKey.contains($0) }) { score += 2 }

    // Mirrors the JS reference exactly: a word in the title *starting* with the
    // act's name (loose on purpose — "Artist - Title" uploads come in every
    // punctuation shape), or an explicit "act - title" credit in the haystack.
    for act in knownActs {
      if titleHay.contains(" \(act)") || hay.contains("\(act) - ") {
        score += 4
        let tag = "\(act) (known lofi act)"
        if !hits.contains(tag) { hits.append(tag) }
        break
      }
    }

    // Long-form mixes and 24/7 streams are lofi's habitat; sub-90s clips are not.
    if durationSeconds > 3600 { score += 0.75 }
    else if durationSeconds > 0 && durationSeconds < 90 { score -= 0.75 }
    else if durationSeconds == 0 { score += 0.25 } // live

    let foundMoods = moods.filter { mood in mood.words.contains { hay.contains($0) } }.map(\.name)

    return GateVerdict(
      lofi: score >= threshold,
      score: (score * 100).rounded() / 100,
      threshold: threshold,
      hits: hits,
      penalties: penalties,
      moods: Array(Set(foundMoods)).sorted()
    )
  }

  /// Tags = whatever the uploader wrote, then whatever the gate inferred.
  static func deriveTags(for track: LofiTrack, gate: GateVerdict) -> [String] {
    var out: [String] = []
    let explicit = track.tags.map { $0.replacingOccurrences(of: "#", with: "").trimmingCharacters(in: .whitespaces) }
    for candidate in explicit + gate.hits + gate.moods {
      let clean = candidate.lowercased().trimmingCharacters(in: .whitespaces)
      guard clean.count > 1, !out.contains(clean) else { continue }
      out.append(clean)
      if out.count >= 12 { break }
    }
    return out
  }

  static func deriveMood(for track: LofiTrack, gate: GateVerdict) -> String {
    if let mood = track.mood, !mood.isEmpty { return mood }
    return gate.moods.first ?? "chill"
  }

  /// Attach a verdict + derived tags to a raw track.
  static func stamped(_ track: LofiTrack) -> LofiTrack {
    var t = track
    let gate = evaluate(
      title: t.title,
      description: t.bestDescription,
      tags: t.tags,
      channelName: t.channelName,
      durationSeconds: t.durationSeconds
    )
    t.gate = gate
    t.tags = deriveTags(for: t, gate: gate)
    t.mood = deriveMood(for: t, gate: gate)
    return t
  }
}

File: prototype/src/lofi-filter.js
// ---------------------------------------------------------------------------
// lofi-filter.js — "only youtube lofi stuff".
//
// A genre gate. Every candidate video (from the seed corpus or a live API) is
// scored on its title, description, hashtags, channel and length; anything
// below the threshold never reaches the deck. Same weights and threshold are
// mirrored in ios/LofiGlass/Services/LofiFilter.swift, and tools/gate-check.mjs
// is the executable reference for both.
// ---------------------------------------------------------------------------

// term (regex source), weight, tag contributed when it matches
export const POSITIVE = [
  ["lo[.\\-\\s]?fi", 5, "lofi"],
  ["l o f i", 4, "lofi"],
  ["ｌｏｆｉ", 4, "lofi"],
  ["chill ?hop", 4, "chillhop"],
  ["jazz ?hop", 4, "jazzy"],
  ["(music|beats|sounds|tunes|radio)\\s*(to|for)\\s+(relax|study|sleep|focus|chill|work|drive|code)", 3.5, "beats to …"],
  ["beats? ?to ?(relax|study|sleep|chill|focus|drive)", 4, "beats to …"],
  ["study beats", 3, "study"],
  ["instrumental hip ?hop", 3, "instrumental"],
  ["hip ?hop radio", 3, "radio"],
  ["boom ?bap", 3, "boom bap"],
  ["tape hiss|tape loop|4th ?gen tape", 2.5, "tape"],
  ["\\bvhs\\b|crt|scan ?line", 2, "vhs"],
  ["chill beats", 2, "chill"],
  ["sleep lofi", 3, "sleep"],
  ["anime (edit|loop|amv)", 2, "anime edit"],
  ["type beat", 2, "type beat"],
  ["neo soul", 1.5, "neo soul"],
  ["\\bjazzy\\b", 1.5, "jazzy"],
  ["24\\s*/\\s*7", 2, "24/7"],
  ["lofi girl|chilledcow", 3, "lofi girl"],
  ["\\b(?:prod\\.|beat)\\s*(?:by\\s*)?[a-z0-9_.-]+\\b", 1, "beatmaker"],
];

// Terms that mean "this is not the genre you asked for"
export const NEGATIVE = [
  ["deep house", -4, "deep house"],
  ["progressive house", -4, "progressive house"],
  ["\btrance\b", -3.5, "trance"],
  ["hard ?style", -5, "hardstyle"],
  ["phonk", -3, "phonk"],
  ["\bdrill\b", -2, "drill"],
  ["heavy metal|deathcore|metalcore", -5, "metal"],
  ["k[\s\-]?pop", -2, "k-pop"],
  ["binaural", -3, "binaural beats"],
  ["subliminal|affirmation|solfeggio|\d{3} ?hz", -3.5, "frequencies/affirmations"],
  ["karaoke", -3, "karaoke"],
  ["workout|gym motivation", -3, "workout"],
  ["edm (festival|mix)", -4, "edm"],
  ["lofi (fake|filter scam)", -5, "lofi-bait"],
];

// Channels whose catalogue is curated lofi → trust boost.
export const TRUSTED_CHANNELS = [
  "lofi girl",
  "chilledcow",
  "lofi records",
  "chillhop music",
  "settle",
  "the bootleg boy",
  "afro lofi",
  "jeez",
  "lofi coffee",
  "mimi lofi chill",
  "the japanese town",
  "a lofi soul",
  "lofi shop 24h",
  "flux.fm",
  "lofi corners",
];

// Producers that only exist in this scene: a bare "Artist - Title" upload from
// one of them is lofi even when the description says nothing.
export const LOFI_ACTS = [
  "kudasai", "no spirit", "tonion", "nymano", "yasumu", "hm surf", "lilac", "trxxshed", "jhove",
  "blurred figures", "another silent weekend", "swiftly", "hazue", "noji", "sutton", "thymes",
  "home grown", "luella", "idealism", "sleep bean", "jinsang", "fantompower", "powfu", "dj hazel",
  "luvlee", "mnts", "tenncoats", "cwrd", "philo", "kuun", "vanyforce", "dthcheese", "shimza", "purrple cat",
];

export const MOODS = [
  ["sleep", ["sleep", "bedtime", "dream", "insomnia", "8 hours", "\ud83d\udca4"]],
  ["study", ["study", "exam", "homework", "revision", "relax/study"]],
  ["rain", ["rain", "storm", "thunder", "wet"]],
  ["cafe", ["cafe", "coffee", "barista", "latte"]],
  ["night-drive", ["night drive", "headlights", "midnight", "\uff5e drive", "drive to"]],
  ["jazzy", ["jazz", "sax", "piano trio", "swing"]],
  ["sad", ["sad", "cry", "lonely", "heartbreak", "melanch", "\ud83d\udc94"]],
  ["anime", ["anime", "ghibli", "spirited away", "opening"]],
  ["morning", ["morning", "sunrise", "breakfast", "upbeat"]],
  ["code", ["code", "coding", "programming"]],
  ["focus", ["focus", "deep work", "concentration"]],
];

const norm = (s = "") => ` ${String(s).toLowerCase().replace(/[|·—–]/g, " ").replace(/\s+/g, " ")} `;
const rxEsc = (s) => s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");

/**
 * @returns {{lofi:boolean, score:number, threshold:number, hits:string[], penalties:string[], moods:string[]}}
 */
export function gateTrack({ title = "", description = "", tags = [], channelName = "", durationSeconds = 0 } = {}) {
  const hay = norm(title) + "|" + norm(description) + "|" + norm((tags || []).join(" ")) + "|" + norm(channelName);
  const titleHay = norm(title);
  let score = 0;
  const hits = [];
  const penalties = [];

  for (const [term, weight, tag] of POSITIVE) {
    const re = new RegExp(term, "g");
    const inTitle = re.test(titleHay);
    re.lastIndex = 0;
    if (inTitle || re.test(hay)) {
      score += weight + (inTitle ? 1 : 0);
      hits.push(tag || term);
    }
  }

  for (const [term, weight, label] of NEGATIVE) {
    if (new RegExp(term, "g").test(hay)) {
      score += weight;
      penalties.push(label || term);
    }
  }

  if (TRUSTED_CHANNELS.some((c) => norm(channelName).includes(rxEsc(c)))) score += 2;
  for (const act of LOFI_ACTS) {
    if (titleHay.includes(` ${act} `) || titleHay.includes(` ${act}`) || norm(`${title} ${description}`).includes(`${act} - `)) {
      score += 4;
      hits.push(`${act} (known lofi act)`);
      break;
    }
  }

  // Long-form mixes and streams are lofi's natural habitat; sub-90s clips are not.
  if (durationSeconds > 3600) score += 0.75;
  else if (durationSeconds > 0 && durationSeconds < 90) score -= 0.75;
  if (durationSeconds === 0) score += 0.25; // live

  const threshold = 4;
  const moods = [];
  for (const [mood, words] of MOODS) if (words.some((w) => hay.includes(w))) moods.push(mood);

  return {
    lofi: score >= threshold,
    score: +score.toFixed(2),
    threshold,
    hits: [...new Set(hits)],
    penalties: [...new Set(penalties)],
    moods: [...new Set(moods)],
  };
}

/** Unique tags: explicit hashtags first, then everything the gate inferred. */
export function deriveTags(track, gate) {
  const explicit = (track.tags || []).map((t) => String(t).replace(/^#/, "").trim()).filter(Boolean);
  const inferred = (gate?.hits || []).concat(gate?.moods || []);
  const out = [];
  for (const t of [...explicit, ...inferred]) {
    const clean = t.replace(/^#/, "").trim().toLowerCase();
    if (clean && clean.length > 1 && !out.includes(clean)) out.push(clean);
    if (out.length >= 12) break;
  }
  return out;
}

export function deriveMood(track, gate) {
  if (track.mood) return track.mood;
  return gate?.moods?.[0] || "chill";
}
